File: src/alchemy.rs
pub const LIQUIDS: [&str; 30] = ["acid","alcohol","blood",
    "blood_fungi","blood_worm","cement",
    "lava","berserkium","pheromone",
    "levitatium","hastium","invisiblium",
    "concentrated_mana","acceleratium","ambrosia",
    "teleportatium","unstable_polymorphine","unstable_teleportatium",
    "worm_pheromone","flummoxium","mud",
    "oil","poison","toxic_sludge",
    "swamp","urine","water",
    "water_ice","water_swamp","chaotic_polymorphine"];
pub const SOLIDS: [&str; 18] = ["bone","brass","coal",
    "copper","diamond","fungus",
    "gold","grass","gunpowder",
    "gunpowder_explosive","rotten_meat","sand",
    "silver","slime","snow",
    "soil","wax","honey"];
const I32: i64 = i32::max_value() as i64;
const F32: f64 = i32::max_value() as f64;

//prepare seed for prng function
pub fn init(seed: i64) -> i64 {
    let iseed = (seed as f64 * 0.17127000 + 1323.59030000) as i64;
    let iseed = lgm_random(iseed,6);
    return iseed;
}

//core prng function
pub fn lgm_random(mut iseed: i64, mut count: i64) -> i64 {
    while count > 0 {
        iseed = 16807 * (iseed % 127773) - 2836 * (iseed / 127773);
        if iseed < 0 {
            iseed = iseed + I32;
        };
        count-=1;
    }
    return iseed;
}

//calculate materials and misnomer "probability" value
pub fn recipe(seed: i64, mut iseed: i64) -> ([&'static str; 4], i64, i64) {
    let mut index: [usize; 4] = [255; 4];
    let mut mats: [&str; 4] = [""; 4];
    let mut x = 0;
    while x < 3 {
        iseed = lgm_random(iseed,1);
        let tmp = (iseed as f64 / F32 * LIQUIDS.len() as f64) as usize;
        if !(index.contains(&tmp)) {
            index[x] = tmp;
            x+=1;
        }    //if duplicate liquid, discard and recycle iseed
    }
    iseed = lgm_random(iseed,1);
    index[3] = (iseed as f64 / F32 * SOLIDS.len() as f64) as usize;
    for n in 0..3 {
        mats[n] = LIQUIDS[index[n]];
    }
    mats[3] = SOLIDS[index[3]];
    mats = shuffle(mats,seed);
    iseed = lgm_random(iseed,1);
    let prob = 10 - ((iseed as f64 / F32) as f64 * -91.0) as i64;
    iseed = lgm_random(iseed,1);
    return (mats,prob,iseed);
}

//shuffle materials; material in last position is then omitted from recipe
fn shuffle (mut mats: [&str; 4], seed: i64) -> [&str; 4] {
    let mut nseed = (seed as i64 >> 1) + 12534;
    nseed = lgm_random(nseed,1);
    let mut index: [usize; 3] = [0; 3];
    for n in 0..3 {
        nseed = lgm_random(nseed,1);
        index[n] = (nseed as f64 / F32 * ( (3 - n) as f64 + 1.0)) as usize;
    }
    let mut x: usize = 3;
    for n in 0..3 {
        mats.swap(x, index[n]);
        x-=1;
    }
    return mats;
}

File: docs/alchemy.js
const Liquids = ["Acid","Whiskey (alcohol)","Blood",
  "Fungus Blood","Worm Blood","Cement",
  "Lava","Berserkium","Pheromone",
  "Levitatium","Hastium","Invisiblium",
  "Concentrated Mana","Acceleratium","Ambrosia",
  "Teleportatium","Unstable Polymorphine","Unstable Teleportatium",
  "Worm Pheromone","Flummoxium","Mud",
  "Oil","Poison","Toxic Sludge (radioactive_liquid)",
  "Swamp (swamp)","Urine","Water",
  "Chilly Water (water_ice)","Swamp (water_swamp)","Chaotic Polymorphine"];
const Solids = ["Bone Dust (bone)","Brass","Coal",
  "Copper","Diamond","Fungus (fungi)",
  "Gold","Grass","Gunpowder",
  "Explosive Gunpowder","Rotten Meat","Sand",
  "Silver","Slime","Snow",
  "Soil","Wax","Honey"];

const LiquidsBeta = ["Water","Chilly Water (water_ice)","Swamp Water (water_swamp)",
  "Oil","Whiskey (alcohol)","Swamp (swamp)",
  "Mud","Blood","Fungus (blood_fungi)",
  "Worm Blood","Toxic Sludge (radioactive_liquid)","Cement",
  "Acid","Lava","Urine",
  "Poison (glowing_liquid)","Teleportatium","Polymorphine",
  "Chaotic Polymorphine","Berserkium","Pheromone",
  "Invisiblium"];
const SolidsBeta = ["Sand","Bone Dust (bone)","Soil",
  "Honey","Slime","Snow",
  "Rotten Meat","Wax","Gold",
  "Silver","Copper","Brass",
  "Diamond","Coal","Gunpowder",
  "Explosive Gunpowder","Grass","Fungus (fungi)"];

const I32 = 2147483647;
const SEEDMAX = 4294967295;

function main (seed) {
  var iseed = init(seed);
  var shuf = initShuffle(seed);
  var tmp = recipe(iseed,shuf);
  return tmp;
}

//prepares rng 'iseed' for recipe calculations
function init (seed) {
  var iseed = parseInt(seed * 0.17127000 + 1323.59030000);
  iseed = lgmRandom(iseed,5);
  return iseed;
}

//calculates shuffle indices for later use
function initShuffle (seed) {
  var nseed = (seed >> 1) + 12534;
  nseed = lgmRandom(nseed,1);
  var shuf = [];
  for (n = 0; n < 3; n++) {
    nseed = lgmRandom(nseed,1);
    shuf[n] = parseInt(nseed / I32 * (3 - n + 1));
  }
  return shuf;
}

//rng calculation; takes number and a 'count' var for iterations
function lgmRandom (iseed,count) {
  while (count > 0) {
    iseed = 16807 * (iseed % 127773) - 2836 * parseInt(iseed / 127773);
    if (iseed < 0) {
      iseed = iseed + I32;
    };
    count--;
  }
  return iseed;
}

//pass rng value to materials fn and calculate probability
function recipe(iseed,shuf) {
  var mats = [];
  var i = 0;
  var prob = [];
  for (x = 0; x < 2; x++) {
    iseed = lgmRandom(iseed,1);
    var tmp = materials(iseed,shuf);
    var array = tmp[0];
    iseed = tmp[1];
    for (n = 0; n < 3; n++) {
      mats[i] = array[n];
      i++;
    }
    iseed = lgmRandom(iseed,1);
    prob[x] = 10 - parseInt(iseed / I32 * -91);
  }
  var tmp = [mats,prob];
  return tmp;
}

//select materials from ingredients list by index
function materials(iseed,shuf) {
  var index = [];
  var array = [];
  var i = 0;
  if (oldrecipes.checked == false) {
    while (i < 3) {
      iseed = lgmRandom(iseed,1);
      var tmp = parseInt(iseed / I32 * Liquids.length);
      if (!index.includes(tmp)) {
        index[i] = tmp;
        i++;
      }
    }
    iseed = lgmRandom(iseed,1);
    index[3] = parseInt(iseed / I32 * Solids.length);
    for (n = 0; n < 3; n++) {
      array[n] = Liquids[index[n]];
    }
    array[3] = Solids[index[3]];
  } else {
    while (i < 3) {
      iseed = lgmRandom(iseed,1);
      var tmp = parseInt(iseed / I32 * LiquidsBeta.length);
      if (!index.includes(tmp)) {
        index[i] = tmp;
        i++;
      }
    }
    iseed = lgmRandom(iseed,1);
    index[3] = parseInt(iseed / I32 * SolidsBeta.length);
    for (n = 0; n < 3; n++) {
      array[n] = LiquidsBeta[index[n]];
    }
    array[3] = SolidsBeta[index[3]];
  }
  array = shuffle(array,shuf);
  var index = [array,iseed];
  return index;
}

//shuffle array with shuffle indices calculated earlier
function shuffle(array,shuf) {
  var i = 3;
  for (n = 0; n < 3; n++) {
    if (shuf[n] != i) {
      [array[i], array[shuf[n]]] = [array[shuf[n]], array[i]];
    }
    i--;
  }
  return array;
}

//print to console
function print(seed,tmp) {
  var mats = tmp[0];
  var prob = tmp[1];
  console.log("\nSeed:",seed);
  console.log("Lively Concoction:",mats[0],mats[1],mats[2]);
  console.log("Alchemic Precursor:",mats[3],mats[4],mats[5]);
  console.log("Lively Concoction Probability:",prob[0]+"%;","Alchemic Precursor Probability:",prob[1]+"%");
}
